src/engine/Filter.cpp

```diff
@@ -45,78 +44,104 @@ string Filter::getDescriptor() const {
 // _____________________________________________________________________________
 ProtoResult Filter::computeResult(bool requestLaziness) {
   LOG(DEBUG) << "Getting sub-result for Filter result computation..." << endl;
-  std::shared_ptr<const Result> subRes = _subtree->getResult(requestLaziness);
+  std::shared_ptr<const Result> subRes = _subtree->getResult(true);
   LOG(DEBUG) << "Filter result computation..." << endl;
   checkCancellation();
 
+  auto localVocab = subRes->getSharedLocalVocab();
   if (subRes->isFullyMaterialized()) {
     IdTable result = filterIdTable(subRes, subRes->idTable());
     LOG(DEBUG) << "Filter result computation done." << endl;
 
-    return {std::move(result), resultSortedOn(), subRes->getSharedLocalVocab()};
+    return {std::move(result), resultSortedOn(), std::move(localVocab)};
   }
-  auto localVocab = subRes->getSharedLocalVocab();
-  return {[](auto subRes, auto* self) -> cppcoro::generator<IdTable> {
-            for (IdTable& idTable : subRes->idTables()) {
-              IdTable result = self->filterIdTable(subRes, idTable);
-              co_yield result;
-            }
-          }(std::move(subRes), this),
-          resultSortedOn(), std::move(localVocab)};
+  if (requestLaziness) {
+    return {[](auto subRes, auto* self) -> cppcoro::generator<IdTable> {
+              for (IdTable& idTable : subRes->idTables()) {
+                IdTable result = self->filterIdTable(subRes, idTable);
+                co_yield result;
+              }
+            }(std::move(subRes), this),
+            resultSortedOn(), std::move(localVocab)};
+  }
+
```

Comment:
Add an `AD_CORRECTNESS_CHECK(subRes lazy, but we materialize)` here, to self-document the logical flow here.

Reply:
There's a literal if statement a couple of lines above that checks exactly that with a return statement.
So I'll just go with a comment instead